Pinta.Effects/Effects/FeatherEffect.cs

```diff
@@ -0,0 +1,143 @@
+	{
+		chrome = services.GetService<IChromeService> ();
+		system = services.GetService<ISystemService> ();
+		EffectData = new FeatherData ();
+	}
+
+	public override void LaunchConfiguration ()
+		=> chrome.LaunchSimpleEffectDialog (this);
+
+	protected override void Render (ImageSurface src, ImageSurface dest, RectangleI roi)
+	{
+		int bottom = roi.Bottom;
+		int src_height = src.Height;
+		int radius = Data.Radius;
+		int threads = system.RenderThreads;
+
+		ConcurrentBag<PointI> borderPixels = new ConcurrentBag<PointI> ();
+		// Color in any pixel that the stencil says we need to fill
+		// First pass
+		// Clean up dest, then collect all border pixels
+		Parallel.For (0, bottom, new ParallelOptions { MaxDegreeOfParallelism = threads }, y => {
+			var src_data = src.GetReadOnlyPixelData ();
+			int src_width = src.Width;
+			var dst_data = dest.GetPixelData ();
+
+			// reset dest to src
+			// Removing this causes preview to not update to lower radius levels
+			var src_row = src_data.Slice (y * src_width, src_width);
+			var dst_row = dst_data.Slice (y * src_width, src_width);
+			for (int x = roi.Left; x <= roi.Right; x++) {
+				dst_row[x].Bgra = src_row[x].Bgra;
+			}
+
+			Span<PointI> pixels = stackalloc PointI[] { new (0, 0), new (0, 0), new (0, 0), new (0, 0) };
+			// Collect a list of pixels that surround the object (border pixels)
+			for (int x = roi.Left; x <= roi.Right; x++) {
+				PointI potentialBorderPixel = new (x, y);
+				if (Data.FeatherCanvasEdge && (x == 0 || x == src_width - 1 || y == 0 || y == src_height - 1)) {
+					borderPixels.Add (potentialBorderPixel);
+				} else if (src.GetColorBgra (src_data, src_width, potentialBorderPixel).A <= Data.Tolerance) {
+					// Test pixel above, below, left, & right
+					pixels[0] = new (x - 1, y);
+					pixels[1] = new (x + 1, y);
+					pixels[2] = new (x, y - 1);
+					pixels[3] = new (x, y + 1);
+
+					foreach (var pixel in pixels) {
+						var px = pixel.X;
+						var py = pixel.Y;
+						if (px < 0 || px >= src_width || py < 0 || py >= src_height)
+							continue;
+						if (src.GetColorBgra (src_data, src_width, new PointI (px, py)).A > Data.Tolerance) {
+							borderPixels.Add (potentialBorderPixel);
+							// Remove comments below to draw border pixels
+							// You will also have to comment out the feather pass because it will overwrite this
+							//int pos = src_width * y + x;
+							//dst_data[pos].Bgra = 0;
+							//dst_data[pos].A = 255;
+
+							break;
+						}
+					}
+				}
+			}
+		});
+
+		// Second pass
+		// Feather pixels according to distance to border pixels
+		Parallel.For (0, bottom, new ParallelOptions { MaxDegreeOfParallelism = threads }, py => {
+			var src_data = src.GetReadOnlyPixelData ();
+			int src_width = src.Width;
+			var dst_data = dest.GetPixelData ();
+
+			var relevantBorderPixels = borderPixels.Where (borderPixel => borderPixel.Y > py - radius && borderPixel.Y < py + radius).ToArray ();
```

Comment:
A future improvement for another PR would be to set up an acceleration structure for the closest border pixel lookup (e.g. KD tree), to avoid going through all of the border pixels for every row. I think we have a spatial hashing implementation in the shape tools as well